h/views/admin_organizations.py

```diff
@@ -0,0 +1,148 @@
+                                        buttons=(_('Create organization'),))
+
+    @view_config(request_method='GET')
+    def get(self):
+        self.form.set_appstruct({
+            'authority': self.request.authority,
+        })
+        return self._template_context()
+
+    @view_config(request_method='POST')
+    def post(self):
+        def on_success(appstruct):
+            authority = appstruct['authority']
+            logo = appstruct['logo']
+            name = appstruct['name']
+            org = Organization(authority=authority, name=name, logo=logo)
+
+            self.request.db.add(org)
+            self.request.session.flash(
+                Markup('Created new organization {}'.format(name)), 'success')
+
+            return HTTPFound(location=self.request.route_url('admin_organizations'))
+
+        return form.handle_form_submission(self.request, self.form,
+                                           on_success=on_success,
+                                           on_failure=self._template_context)
+
+    def _template_context(self):
+        return {'form': self.form.render()}
+
+
+@view_defaults(route_name='admin_organizations_edit',
+               permission='admin_organizations',
+               renderer='h:templates/admin/organizations_edit.html.jinja2')
+class OrganizationEditController(object):
+
+    def __init__(self, org, request):
+        self.org = org
+        self.request = request
+        self.schema = OrganizationSchema().bind(request=request)
+        self.form = request.create_form(self.schema,
+                                        buttons=(_('Save'),))
+        self._update_appstruct()
+
+    @view_config(request_method='GET')
+    def read(self):
+        return self._template_context()
+
+    @view_config(request_method='POST',
+                 route_name='admin_organizations_delete')
+    def delete(self):
+        org = self.org
+
+        # Prevent deletion while the organization has associated groups.
+        group_count = (self.request.db.query(models.Group)
+                                      .filter_by(organization=org)
+                                      .count())
+        if group_count > 0:
+            self.request.response.status_int = 400
+            self.request.session.flash(
+                _('Cannot delete organization because it is associated with {} groups'
+                  .format(group_count)), 'error')
+            return self._template_context()
+
+        # Delete the organization.
+        self.request.db.delete(org)
+        self.request.session.flash(
+            _('Successfully deleted organization %s' % (org.name), 'success'))
+        return HTTPFound(
+            location=self.request.route_path('admin_organizations'))
+
+    @view_config(request_method='POST')
+    def update(self):
+        org = self.org
+
+        def on_success(appstruct):
+            org.name = appstruct['name']
+            org.logo = appstruct['logo']
```

Comment:
Naivete-with-SQLAlchemy alert! I was looking at the code here trying to spot the place where the db gets updated, but I presume that merely updating/dirtying the model object here will cause the changes to be persisted, correct?

Reply:
Yes. Changes to model objects get committed at the end of the web request (see [my writeup of how this works](https://robertknight.me.uk/posts/pyramid-db-transactions/)) if they have not already been explicitly flushed.

The main reason for having explicit flushes in our code is to get the value of some field (eg. a new group's `pubid` or most objects' `id` fields) which is only generated when the object is saved.